File: Lecture Works/Bland Brewery Production/bland_brewery.m
profit_coefficients = [-13; -23];  

constraint_matrix = [5 15; 4 4; 35 20];   
constraint_rhs = [480; 160; 1190];        

lower_bounds = [0; 0];

[optimal_solution, max_profit] = linprog(profit_coefficients, constraint_matrix, constraint_rhs, [], [], lower_bounds, []);

max_profit = -max_profit;

optimal_ale = optimal_solution(1);
optimal_beer = optimal_solution(2);

fprintf('Optimal number of Ale barrels: %.0f\n', optimal_ale);
fprintf('Optimal number of Beer barrels: %.0f\n', optimal_beer);
fprintf('Maximum profit: $%.2f\n', max_profit);

ale_barrels = linspace(0, 100, 500);  

corn_constraint_beer = (480 - 5 * ale_barrels) / 15;
hops_constraint_beer = (160 - 4 * ale_barrels) / 4;
malt_constraint_beer = (1190 - 35 * ale_barrels) / 20;

figure('Position', [100, 100, 800, 600], 'Color', 'w');
hold on;

plot(ale_barrels, corn_constraint_beer, 'r', 'LineWidth', 2, 'DisplayName', 'Corn Capacity');
plot(ale_barrels, hops_constraint_beer, 'y', 'LineWidth', 2, 'DisplayName', 'Hops Capacity');
plot(ale_barrels, malt_constraint_beer, 'b', 'LineWidth', 2, 'DisplayName', 'Malt Capacity');

feasible_beer = min([corn_constraint_beer; hops_constraint_beer; malt_constraint_beer]);

fill([ale_barrels, fliplr(ale_barrels)], [feasible_beer, zeros(1, 500)], 'g', 'FaceAlpha', 0.2, 'DisplayName', 'Feasible Region');

plot(optimal_ale, optimal_beer, 'r.', 'MarkerSize', 25, 'DisplayName', sprintf('Optimal Solution (Ale = %.0f, Beer = %.0f)', optimal_ale, optimal_beer));

xlabel('Barrels of Ale', 'FontSize', 12);
ylabel('Barrels of Beer', 'FontSize', 12);
title('Bland Brewery Production Possibility Frontier', 'FontSize', 14);

xlim([0 100]);
ylim([0 60]);

set(gca, 'XTick', 0:10:100, 'YTick', 0:10:60);
set(gca, 'XMinorTick', 'on', 'YMinorTick', 'on');
set(gca, 'XMinorGrid', 'on', 'YMinorGrid', 'on');

grid on;

legend('show', 'Location', 'Best');
hold off;

File: Lecture Works/Bland Brewery Production/bland_brewery.py
import numpy as np
import matplotlib.pyplot as plt
import pulp

bland_brewery = pulp.LpProblem('Bland_Brewery', pulp.LpMaximize)

ale = pulp.LpVariable('Barrels of Ale', lowBound=0, cat='Integer')
beer = pulp.LpVariable('Barrels of Beer', lowBound=0, cat='Integer')

bland_brewery += 13 * ale + 23 * beer, 'Total Profit'

bland_brewery += 5 * ale + 15 * beer <= 480, 'Corn Capacity'
bland_brewery += 4 * ale + 4 * beer <= 160, 'Hops Capacity'
bland_brewery += 35 * ale + 20 * beer <= 1190, 'Malt Capacity'

bland_brewery.solve()

optimal_ale = pulp.value(ale)
optimal_beer = pulp.value(beer)
optimal_profit = pulp.value(bland_brewery.objective)

print(f'Optimal number of Ale barrels: {optimal_ale}')
print(f'Optimal number of Beer barrels: {optimal_beer}')
print(f'Maximum profit: {optimal_profit}')

corn_constraint_ale = np.linspace(0, 100, 500)
corn_constraint_beer = (480 - 5 * corn_constraint_ale) / 15
hops_constraint_ale = np.linspace(0, 100, 500)
hops_constraint_beer = (160 - 4 * hops_constraint_ale) / 4
malt_constraint_ale = np.linspace(0, 100, 500)
malt_constraint_beer = (1190 - 35 * malt_constraint_ale) / 20

plt.figure(figsize=(10, 7), dpi=300, facecolor='w', edgecolor='w', clear=True)

plt.plot(corn_constraint_ale, corn_constraint_beer, label='Corn', color='red')
plt.plot(hops_constraint_ale, hops_constraint_beer, label='Hops', color='yellow')
plt.plot(malt_constraint_ale, malt_constraint_beer, label='Malt', color='blue')

plt.fill_between(corn_constraint_ale, 0, corn_constraint_beer, where=(corn_constraint_beer >= 0), alpha=0.1, color='red')
plt.fill_between(hops_constraint_ale, 0, hops_constraint_beer, where=(hops_constraint_beer >= 0), alpha=0.1, color='yellow')
plt.fill_between(malt_constraint_ale, 0, malt_constraint_beer, where=(malt_constraint_beer >= 0), alpha=0.1, color='blue')

plt.xlim(0, 100)
plt.ylim(0, 60)

plt.xlabel('Barrels of Ale')
plt.ylabel('Barrels of Beer')
plt.title('Bland Brewery Production Possibility Frontier')

plt.xticks(np.arange(0, 101, 10))
plt.yticks(np.arange(0, 61, 10))
plt.minorticks_on()
plt.gca().xaxis.set_minor_locator(plt.MultipleLocator(1))
plt.gca().yaxis.set_minor_locator(plt.MultipleLocator(1))
plt.grid(which='both', axis='both', linestyle='solid', linewidth=0.5)

plt.plot(optimal_ale, optimal_beer, 'ro', label=f'Optimal Solution (Ale = {optimal_ale:.0f}, Beer = {optimal_beer:.0f})')

plt.grid(True)
plt.legend()
plt.savefig('Bland Brewery.png')
plt.show()
